pkg/network/server.go

```diff
@@ -620,11 +620,30 @@ func (s *Server) handleGetRootsCmd(p Peer, gr *payload.GetStateRoots) error {
 }
 
 // handleStateRootsCmd processees `roots` request.
-func (s *Server) handleRootsCmd(rs *payload.StateRoots) error {
+func (s *Server) handleRootsCmd(p Peer, rs *payload.StateRoots) error {
 	for i := range rs.Roots {
 		_ = s.chain.AddStateRoot(&rs.Roots[i])
 	}
-	return nil
+	// request more state roots from peer if needed
+	return s.requestStateRoot(p)
+}
+
+// requestStateRoot sends `getroots` message to get verified state roots.
+func (s *Server) requestStateRoot(p Peer) error {
+	stateHeight := s.chain.StateHeight()
+	hdrHeight := s.chain.HeaderHeight()
```

Comment:
It should be block height, I think. You have to have the state to verify it and with headers only you don't yet have it.

Reply:
Fixed.